src/psm/src/ir_network.cpp

```diff
@@ -0,0 +1,1299 @@
+
+  if (box->isVia()) {
+    // handle as via
+    std::vector<odb::dbShape> via_shapes;
+    box->getViaBoxes(via_shapes);
+
+    for (const auto& shape : via_shapes) {
+      auto* layer = shape.getTechLayer();
+      if (layer->getRoutingLevel() == 0) {
+        // via box
+      } else {
+        // enclosure box
+        Polygon90Set& shapes = shapes_by_layer[layer];
+
+        odb::Rect rect = shape.getBox();
+        transform.apply(rect);
+
+        shapes += rectToPolygon(rect);
+      }
+    }
+  } else {
+    // handle as shape
+    auto* layer = box->getTechLayer();
+
+    odb::Rect rect = box->getBox();
+    transform.apply(rect);
+    shapes_by_layer[layer] += rectToPolygon(rect);
+  }
+
+  return shapes_by_layer;
+}
+
+IRNetwork::LayerMap<IRNetwork::Polygon90Set>
+IRNetwork::generatePolygonsFromSWire(odb::dbSWire* wire)
+{
+  using boost::polygon::operators::operator+=;
+
+  const utl::DebugScopedTimer timer(
+      logger_, utl::PSM, "timer", 1, "Generate shapes from SWire: {}");
+
+  LayerMap<Polygon90Set> shapes_by_layer;
+
+  for (odb::dbSBox* box : wire->getWires()) {
+    for (const auto& [layer, polygon] :
+         generatePolygonsFromBox(box, odb::dbTransform())) {
+      shapes_by_layer[layer] += polygon;
+    }
+  }
+
+  return shapes_by_layer;
+}
+
+IRNetwork::LayerMap<IRNetwork::Polygon90Set>
+IRNetwork::generatePolygonsFromITerms(std::vector<TerminalNode*>& terminals)
+{
+  using boost::polygon::operators::operator+=;
+
+  const utl::DebugScopedTimer timer(
+      logger_, utl::PSM, "timer", 1, "Generate shapes from ITerms: {}");
+
+  auto* base_layer = getTech()->findRoutingLayer(1);
+
+  LayerMap<Polygon90Set> shapes_by_layer;
+  for (auto* iterm : net_->getITerms()) {
+    auto* inst = iterm->getInst();
+    if (!inst->isPlaced()) {
+      continue;
+    }
+    if (floorplanning_ && !inst->getPlacementStatus().isFixed()) {
+      continue;
+    }
+    const auto transform = inst->getTransform();
+    int x, y;
+    iterm->getAvgXY(&x, &y);
+    auto base_node
+        = std::make_unique<ITermNode>(iterm, odb::Point(x, y), base_layer);
+
+    bool has_routing_term = false;
+    for (auto* mpin : iterm->getMTerm()->getMPins()) {
+      for (auto* geom : mpin->getGeometry()) {
+        for (const auto& [layer, shapes] :
+             generatePolygonsFromBox(geom, transform)) {
+          shapes_by_layer[layer] += shapes;
+        }
+
+        if (geom->isVia()) {
+          continue;
+        }
+
+        auto* layer = geom->getTechLayer();
+
+        has_routing_term = true;
+
+        odb::Rect pin_shape = geom->getBox();
+        transform.apply(pin_shape);
+
+        shapes_by_layer[layer] += rectToPolygon(pin_shape);
+
+        // create iterm nodes
+        auto center = std::make_unique<TerminalNode>(pin_shape, layer);
+        terminals.push_back(center.get());
+
+        connections_.push_back(
+            std::make_unique<TermConnection>(base_node.get(), center.get()));
```

Comment:
If the base_node is at the same location as the center then can you skip making the center node and the connection?  For simple pins this would often occur.

Reply:
Same with generatePolygonsFromBTerms

Reply:
The `base_node` is special (only one is created per ITerm and BPin), to avoid a bunch of logic to prevent merging the iterm and bpin nodes away, they are kept separate from the rest of the nodes, but this requires a few extra connections to track them.